### PR TITLE
KB article for NetApp Trident CSI Driver

### PR DESCRIPTION
**Pull request by jeff-radick-suse**
Add knowledge base article describing how to install the NetApp Astra Trident CSI driver to enable use of NetApp storage systems for Harvester volume storage.

### CODE REVIEW

**Comment by w13915984028** [comment]
FYI:

4a, 4b ... are not well formated
image

5a is similar
image

**Comment by jeff-radick-suse** [comment]
Thank you for the formatting comments.  After some feedback received by direct communication, some of the steps in the document will be removed and replaced with an instruction to refer to NetApp Trident documentation.  This includes much of the badly formatted parts of the document.

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+== Detailed Instructions
+
+We are assuming that before beginning this procedure,
+a Harvester cluster and a NetApp ONTAP storage system
+are both installed and configured for use.
+
+Follow all of the following steps, but ONLY on the initial cluster node.
+
+The last step (setting up multipathd) should be done on all nodes,
+AFTER the initial node has had Trident installed.
```
Comment (ibrokethecloud) [inline kb/2023-08-11/using_netapp_third_party_storage.md:38]: we should rephrase this to may be `After the Trident CSI has been installed` it does not matter how many nodes the cluster already has.

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+1. Read the NetApp Astra Trident documentation
+       https://docs.netapp.com/us-en/trident/
+       https://docs.netapp.com/us-en/trident/trident-get-started/kubernetes-deploy-operator.html
+   We are describing the "manual deployment" procedure here.
+2. Log into the Harvester cluster in one of the nodes.
```
Comment (ibrokethecloud) [inline kb/2023-08-11/using_netapp_third_party_storage.md:44]: users need to login to one of the `master` nodes or download the kubeconfig and run the commands locally.

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    Installed 67s trident-operator.netapp.io Trident installed
+```
+
+4.d. Verify the installation
+   - `kubectl get pods -n trident`
+   - `./tridentctl -n trident version`
+
+5. Install a backend.
+   This part is specific to Harvester.
+5.a.    Put the following into a text file, for example /tmp/backend.yaml
```
Comment (ibrokethecloud) [inline kb/2023-08-11/using_netapp_third_party_storage.md:83]: we should may be add some explanation for the fields in the backend.yaml? or also point to the official spec?

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+```
+7.b. Enable multipathd
+    The above file will take effect on the next reboot of the node;
+    multipathd can be enabled immediately without rebooting the node
+    using the following commands:
+    - `systemctl enable multipathd`
+    - `systemctl start multipathd`
+
+After the above steps, the `ontap-san-economy` storage class
+should be available for use when creating a volume for
```
Comment (ibrokethecloud) [inline kb/2023-08-11/using_netapp_third_party_storage.md:153]: can we please add info on either disabling mutlipathd for longhorn or additional steps for disabling multipathd support in ontap, plus reasons for the same.

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+We are assuming that before beginning this procedure,
+a Harvester cluster and a NetApp ONTAP storage system
+are both installed and configured for use.
+
+Most of these steps can be performed on any system that has the `helm` and `kubectl` commands installed, and which has network connectivity to the management port of the Harvester cluster.  Let's call this your workstation.  Certain steps must be performed on one or more cluster nodes themselves.  The steps described below should be done on your workstation unless otherwise indicated.
+
+The last step (setting up multipathd) should be done on all nodes,
+after the Trident CSI has been installed.
+
+Certain parameters of your installation will require modification of details in the examples in the procedure given below. This which you may wish to modify include:
```
Comment (markhillgit) [inline kb/2023-08-11/using_netapp_third_party_storage.md:37]: perhaps "Those which you may wish to modify"?

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+1. Install the helm chart.
+
+   * First you will need to add the Astra Trident Helm repository:
+
+      ```shell
+      helm repo add netapp-trident https://netapp.github.io/trident-helm-chart
+      ```
+
+   * Next, install the Helm cart.  This example uses `mytrident` as the deployment name, `trident` as the namespace, and 23.07.0 as the version number to install:
```
Comment (markhillgit) [inline kb/2023-08-11/using_netapp_third_party_storage.md:105]: "helm chart" not cart

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 5 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:37]: right

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 6 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:105]: right

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 4 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md]: I think I've done this now.  We've discussed what I've done but any comments on the results will be appreciated

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 3 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md]: I didn't think of this.  I will have to look up the right doc.  At the moment, this part is magic to me.

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 2 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md]: I think in my major revision I've described the method for downloading and using the kubeconfig locally, along with the procedure for using helm.  Your feedback on those changes will be appreciated.

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 1 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md]: Done

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 2 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md]: I didn't say anything about doing stuff on a `master` node though.  Perhaps I should?

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+3rd party storage classes (including those based on Trident) can only be used for non-boot volumes of Harvester VMs.
+:::
+
+# Detailed Instructions
+
+We are assuming that before beginning this procedure,
+a Harvester cluster and a NetApp ONTAP storage system
+are both installed and configured for use.
+
+Most of these steps can be performed on any system that has the `helm` and `kubectl` commands installed, and which has network connectivity to the management port of the Harvester cluster.  Let's call this your workstation.  Certain steps must be performed on one or more cluster nodes themselves.  The steps described below should be done on your workstation unless otherwise indicated.
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:32]:
```suggestion
Most of these steps can be performed on any system with the `helm` and `kubectl` commands installed and network connectivity to the management port of the Harvester cluster. Let's call this your workstation. Certain steps must be performed on one or more cluster nodes themselves. The steps described below should be done on your workstation unless otherwise indicated.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+The last step (setting up multipathd) should be done on all nodes,
+after the Trident CSI has been installed.
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:35]:
```suggestion
The last step (enabling multipathd) should be done on all nodes after the Trident CSI has been installed.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+
+Certain parameters of your installation will require modification of details in the examples in the procedure given below. Those which you may wish to modify include:
+
+* The namespace.  `trident` is used as the namespace in the examples, but you may prefer to use another.
+* The name of the deployment. `mytrident` is used but you can change this to something else.
+* The management IP address of the ONTAP storage system
+* Login credentials (username and password) of the ONTAP storage system
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:42]:
```suggestion
Specific installation parameters will require modification of details in the examples in the procedure given below. Those which you may wish to modify include:

* The namespace. `trident` is used as the namespace in the examples, but you may prefer to use another.
* The deployment's name. `mytrident` is used, but you can change this to something else.
* The management IP address of the ONTAP storage system.
* Login credentials (username and password) of the ONTAP storage system.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+The procedure is as follows.
+
+1. Read the NetApp Astra Trident documentation:
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:46]:
```suggestion
1. Read the following NetApp Astra Trident documentation:
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+   * https://docs.netapp.com/us-en/trident/trident-get-started/kubernetes-deploy-helm.html#deploy-the-trident-operator-and-install-astra-trident-using-helm
+
+   The simplest method is to install using Helm; that process is described here.
+
+1. Download the KubeConfig from the Harvester cluster.
+
+   * Open the web UI for your Harvester cluster
+   * In the lower left corner, click the "Support" link.  This will take you to a "Harvester Support" page.
+   * Click the button labeled "Download KubeConfig".  This will download a your cluster config in a file called "local.yaml" by default.
+   * Move this file to a convenient location and set your `KUBECONFIG` environment variable to the path of this file.
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:59]:
```suggestion
1. Download the KubeConfig from the Harvester cluster.

   1. Open the web UI for your Harvester cluster.
   1. In the lower-left corner, select the **Support** link. This will take you to the **Harvester Support** page.
   1. Select "Download KubeConfig". This will download your cluster's config file, named `local.yaml` by default.
   1. Move this file to a convenient location and set your `KUBECONFIG` environment variable to the path of this file.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+1. Prepare the cluster for installation of the Helm chart.
+
+   Before starting installation of the helm chart, special authorization must be provided to enable certain modifications to be made during the installation.
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:63]:
```suggestion
1. Prepare the cluster for installation of the Helm chart.

   Before starting the installation of the helm chart, you must provide special authorization to enable certain modifications to be made during the installation.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+
+   * Put the following text into a file.  For this example we'll call it `authorize_trident.yaml`.
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:65]:
```suggestion
   1. Put the following text into a file. For this example, we'll call it `authorize_trident.yaml`.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+        apiGroup: rbac.authorization.k8s.io
+        kind: ClusterRole
+        name: trident-operator-psa
+      subjects:
+      - kind: ServiceAccount
+        name: trident-operator
+        namespace: trident
+      ```
+
+   * Apply this manifest via the command `kubectl apply -f authorize_trident.yaml`.
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:95]:
```suggestion
   1. Apply this manifest via the command `kubectl apply -f authorize_trident.yaml`.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+
+1. Install the helm chart.
+
+   * First you will need to add the Astra Trident Helm repository:
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:99]:
```suggestion
   1. First you will need to add the Astra Trident Helm repository:
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+
+      ```shell
+      helm repo add netapp-trident https://netapp.github.io/trident-helm-chart
+      ```
+
+   * Next, install the Helm chart.  This example uses `mytrident` as the deployment name, `trident` as the namespace, and 23.07.0 as the version number to install:
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:105]:
```suggestion
   1. Next, install the Helm chart. This example uses `mytrident` as the deployment name, `trident` as the namespace, and `23.07.0` as the version number to install:
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+
+      ```shell
+      helm install mytrident netapp-trident/trident-operator --version 23.07.0 --create-namespace --namespace trident
+      ```
+
+   * The NetApp documentation describes variations on how you can do this.
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:111]:
```suggestion
   The NetApp documentation describes variations on how you can do this.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+
+1. Download and extract the tridentctl command, which will be needed for the next few steps:
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:113]:
```suggestion
1. Download and extract the tridentctl command, which is needed for the next few steps:
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+      password: password1234
+      labels:
+      name: default_backend_san
+      ```
+
+    The LIF IP address, username, and password of this file
+    should be replaced with the management LIF and credentials
+    for the ONTAP system.
+
+   1. Create the backend
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:144]:
```suggestion
   1. Create the backend.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+
+      ```shell
+      ./tridentctl create backend -f /tmp/backend.yaml -n trident
+      ```
+
+   1. Check that it is created
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:150]:
```suggestion
   1. Check that it is created.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 21]
```diff
@@ -0,0 +1,281 @@
...
+
+      ```shell
+      ./tridentctl get backend -n trident
+      ```
+
+1. Define a StorageClass and SnapshotClass.
+
+   1. Put the following into a file, for example `/tmp/storage.yaml`
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:158]:
```suggestion
   1. Put the following into a file, for example `/tmp/storage.yaml`.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+      ---
+      apiVersion: snapshot.storage.k8s.io/v1
+      kind: VolumeSnapshotClass
+      metadata:
+      name: csi-snapclass
+      driver: csi.trident.netapp.io
+      deletionPolicy: Delete
+      ```
+
+   1. Apply the definitions:
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:178]:
```suggestion
   1. Apply the definitions.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+
+      ```shell
+      kubectl apply -f /tmp/storage.yaml
+      ```
+
+1. Enable multipathd
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:184]:
```suggestion
1. Enable multipathd. The following is required to enable multipathd. This must be done on every node of the Harvester cluster. The preceding steps should only be done once on a single node.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 24]
```diff
@@ -0,0 +1,281 @@
...
+
+   The following is required to enable multipathd.
+   This must be done on every node of the Harvester cluster.
+   The preceding steps should only be done once on a single node.
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:188]:
```suggestion
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 25]
```diff
@@ -0,0 +1,281 @@
...
+      will not work correctly if `multipathd` tries to use those devices.
+
+      Therefore the file `/etc/multipath.conf` must be set up on each node
+      so as to prevent `multipathd` from touching any of the devices
+      that Longhorn will use.  Unfortunately, it is not possible to know
+      in advance which device names will be used until the volumes are attached
+      to a VM when the VM is started, or when the volumes are hot-added to a running VM.
+      The recommended method is to "whitelist" the Trident devices using device
+      properties rather than device naming.  The properties to allow are the
+      device vendor and product.  Here is an example of what you'll want in `/etc/multipath.conf`:
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:217]:
```suggestion
   1. Configure `multipathd` to exclude pathnames used by Longhorn.

      This part is a little tricky. `multipathd` will automatically discover device names matching a specific pattern and attempt to set up multipathing on them. Unfortunately, Longhorn's device names follow the same pattern and will not work correctly if `multipathd` tries to use those devices.

      Therefore, the file `/etc/multipath.conf` must be set up on each node to prevent `multipathd` from touching any devices that Longhorn will use. Unfortunately, it is not possible to know in advance which device names will be used until the volumes are attached to a VM when the VM is started or when the volumes are hot-added to a running VM. The recommended method is to "whitelist" the Trident devices using device properties rather than device naming. The properties to allow are the device vendor and product. Here is an example of what you'll want in `/etc/multipath.conf`:
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 26]
```diff
@@ -0,0 +1,281 @@
...
+
+       This is only an example that works if NetApp is the only storage provider in the system
+       for which `multipathd` must be used.  More complex environments will require
+       more complex configuration.
+
+       Explicitly putting that content into `/etc/multipath.conf` will work when you start `multipathd` as described below,
+       but the change in `/etc` will not persist across node reboots.  To solve that problem, you should add another
+       file to `/oem` that will re-generate `/etc/multipath.conf` when the node reboots.  The following example
+       will create the `/etc/multipath.conf` given in the example above, but may need to be modified for your
+       environment if you have a more complex iSCSI configuration:
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:242]:
```suggestion
        This example only works if NetApp is the only storage provider in the system for which `multipathd` must be used. More complex environments will require a more complex configuration.

       Explicitly putting that content into `/etc/multipath.conf` will work when you start `multipathd` as described below, but the change in `/etc` will not persist across node reboots. To solve that problem, add another file to `/oem` that will re-generate `/etc/multipath.conf` when the node reboots. The following example will create the `/etc/multipath.conf` given in the example above. Still, it may need to be modified for your environment if you have a more complex iSCSI configuration:
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 27]
```diff
@@ -0,0 +1,281 @@
...
+                   }
+                   blacklist_exceptions {
+                       device {
+                           vendor "NETAPP"
+                           product "LUN"
+                       }
+                   }
+       ```
+
+       Again, this has to be done on every node.
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:268]:
```suggestion
       Remember, this has to be done on every node.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 28]
```diff
@@ -0,0 +1,281 @@
...
+
+   1. Enable multipathd
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:270]:
```suggestion
   1. Enable multipathd.
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 29]
```diff
@@ -0,0 +1,281 @@
...
+
+    Adding the above files to `/oem` will take effect on the next reboot of the node;
+    `multipathd` can be enabled immediately without rebooting the node
+    using the following commands:
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:274]:
```suggestion
    Adding the above files to `/oem` will take effect on the next reboot of the node; `multipathd` can be enabled immediately without rebooting the node using the following commands:
```

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 30]
```diff
@@ -0,0 +1,281 @@
...
+
+      ```shell
+      systemctl enable multipathd
+      systemctl start multipathd
+      ```
+
+      After the above steps, the `ontap-san-economy` storage class should be available for use when creating a volume for a Harvester VM.
```
Comment (LucasSaintarbor) [inline kb/2023-08-11/using_netapp_third_party_storage.md:281]:
```suggestion
      After the above steps, the `ontap-san-economy` storage class should be available when creating a volume for a Harvester VM.
```

**Review by LucasSaintarbor — APPROVED** [review]
Small NITs and formatting changes. LGTM. Please commit these changes if they look okay, thanks!

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 31]
```diff
@@ -0,0 +1,281 @@
...
+description: Installation procedure for NetApp Astra Trident CSI Driver
+slug: install_netapp_trident_csi
+authors:
+  - name: Jeff Radick
+    title: Staff Software Engineer
+tags: [harvester]
+hide_table_of_contents: false
+---
+
+Instructions for installing the Netapp Astra Trident CSI driver into a Harvester cluster.
```
[the remark this answers is not in the document]
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:12]: done

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 32]
```diff
@@ -0,0 +1,281 @@
...
+
+This enables use of NetApp storage systems to store storage volumes usable by virtual machines running in Harvester.
```
[the remark this answers is not in the document]
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:14]: done

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 33]
```diff
@@ -0,0 +1,281 @@
...
+
+:::note
+3rd party storage classes (including those based on Trident) can only be used for non-boot volumes of Harvester VMs.
+:::
+
+# Detailed Instructions
+
+We are assuming that before beginning this procedure,
+a Harvester cluster and a NetApp ONTAP storage system
+are both installed and configured for use.
```
[the remark this answers is not in the document]
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:30]: done

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 7 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:32]: done

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 8 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:35]: done

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 26 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:242]: done

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 27 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:268]: done

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 28 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:270]: done

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 29 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:274]: Done.  I only just realized that some of what you changed before was to join multiple lines of text into a single line, which I guess affects how markdown decides where to put the line breaks.  Going back to adjust all of the previous corrections I'm making to match that change.

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 30 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:281]: done

**Review by markhillgit — APPROVED** [review]

File: kb/2023-08-11/using_netapp_third_party_storage.md [hunk 34]
```diff
@@ -0,0 +1,268 @@
...
+      in advance which device names will be used until the volumes are attached
+      to a VM when the VM is started, or when the volumes are hot-added to a running VM.
+      The recommended method is to "whitelist" the Trident devices using device
+      properties rather than device naming.  The properties to allow are the
+      device vendor and product.  Here is an example of what you'll want in `/etc/multipath.conf`:
+
+       ```text
+       blacklist {
+           device {
+               vendor "!NETAPP"
```
Comment (yasker) [inline kb/2023-08-11/using_netapp_third_party_storage.md:217]: I probably missed something here. What's the reason we decide to whitelist `NetApp` instead of blacklist `Longhorn`?

Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md:217]: I was not able to find any discernible property I could check for in the multipath.conf or even via a script that would positively identify Longhorn devices.  The device names used by Longhorn are of the form `/dev/sd*` but so are the device names used for the NetApp LUNs.  The device names used by Longhorn (and NetApp for that matter) don't even appear in `/dev` until the VM using those volumes gets created, and there was nothing I know how to put in the multipath.conf that would allow me to prevent multipathd from seeing those Longhorn devices.  The only thing I was able to come up with was the device vendor and product strings.  I am not aware of vendor and product strings consistently used by Longhorn.  In my test setup I see a vendor of "IET" (whatever or whoever that is) and a product of "VIRTUAL-DISK" but that might be because my test environment has my Harvester nodes running as VMs on a harvester system, and those strings might be coming from the underlying QEMU/KVM.  I'm not sure.

I could spend more time trying to figure this out, and I may or may not find something better.  For the simple case where NetApp is the only 3rd party storage provider, I know this works, and it seems better to provide a simple working solution than hold out for a perfect universal solution that might not be there.

Comment (yasker) [reply kb/2023-08-11/using_netapp_third_party_storage.md:217]: Ok. This is something Longhorn team should fix. We should be able to identify Longhorn volumes as easy as those NetApp volumes. cc @innobead

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 2 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md]: I actually did put in a few words about this later on.

File: kb/2023-08-11/using_netapp_third_party_storage.md (hunk 5 above)
Comment (jeff-radick-suse) [reply kb/2023-08-11/using_netapp_third_party_storage.md]: done

**Review by ibrokethecloud — APPROVED** [review]
LGTM. thanks.
